db/db_impl.cc

```diff
@@ -1639,6 +1641,271 @@ std::vector<Status> DBImpl::MultiGet(
+          mgd_iter->super_version = GetAndRefSuperVersion(mgd_iter->cfd);
+        } else {
+          mgd_iter->super_version = mgd_iter->cfd->GetSuperVersion()->Ref();
+        }
+        TEST_SYNC_POINT("DBImpl::MultiGet::AfterRefSV");
+        if (read_options.snapshot != nullptr || last_try) {
+          // If user passed a snapshot, then we don't care if a memtable is
+          // sealed or compaction happens because the snapshot would ensure
+          // that older key versions are kept around. If this is the last
+          // retry, then we have the lock so nothing bad can happen
+          continue;
+        }
+        // We could get the earliest sequence number for the whole list of
+        // memtables, which will include immutable memtables as well, but that
+        // might be tricky to maintain in case we decide, in future, to do
+        // memtable compaction.
+        if (!last_try) {
+          auto seq = mgd_iter->super_version->mem->GetEarliestSequenceNumber();
+          if (seq > snapshot) {
+            retry = true;
+            break;
+          }
+        }
+      }
+      if (!retry) {
+        if (last_try) {
+          mutex_.Unlock();
+        }
+        break;
+      }
+    }
+  }
+
+  // Keep track of bytes that we read for statistics-recording later
+  PERF_TIMER_STOP(get_snapshot_time);
+
+  // For each of the given keys, apply the entire "get" process as follows:
+  // First look in the memtable, then in the immutable memtable (if any).
+  // s is both in/out. When in, s could either be OK or MergeInProgress.
+  // merge_operands will contain the sequence of merges in the latter case.
+  for (auto col_iter = multiget_cf_data.begin();
+       col_iter != multiget_cf_data.end(); ++col_iter) {
+    size_t num_keys_in_cf = col_iter->end - col_iter->start;
+    while (num_keys_in_cf) {
+      size_t batch_size = (num_keys_in_cf > MultiGetContext::MAX_KEYS_ON_STACK)
+                              ? MultiGetContext::MAX_KEYS_ON_STACK
+                              : num_keys_in_cf;
+      MultiGetContext ctx(&sorted_keys[col_iter->start], batch_size, snapshot);
+      MultiGetRange range = ctx.GetMultiGetRange();
+      bool lookup_current = false;
+      auto mgd = *col_iter;
+      auto super_version = mgd.super_version;
+
+      num_keys_in_cf -= batch_size;
+      col_iter->start += batch_size;
+      // ctx.InitLookupKeys(snapshot);
+      for (auto mget_iter = range.begin(); mget_iter != range.end();
+           ++mget_iter) {
+        MergeContext& merge_context = mget_iter->merge_context;
+        merge_context.Clear();
+        Status& s = *mget_iter->s;
+        PinnableSlice* value = mget_iter->value;
+        s = Status::OK();
+
+        bool skip_memtable =
+            (read_options.read_tier == kPersistedTier &&
+             has_unpersisted_data_.load(std::memory_order_relaxed));
+        bool done = false;
+        if (!skip_memtable) {
+          if (super_version->mem->Get(
```

Comment:
It's fine for now but do we plan to push down to memtables too?

Reply:
Yes, I think there would be benefit to pushing down to memtable, especially since whole key bloom filter is now supported in memtables. I'll do it in a follow-on PR.